tests/unit/plugins/mesos/retrying_executor_test.py

```diff
@@ -59,37 +73,169 @@ def test_task_retry(mock_retrying_executor):
+    retry_attempted = mock_retrying_executor.retry(mock_event)
 
     assert mock_retrying_executor.task_retries[mock_event.task_id] == 0
-    assert mock_retrying_executor.run.call_count == 1
+    assert mock_retrying_executor.run.call_count == 0
+    assert not retry_attempted
+
+# retry_loop ###################################################################
 
 
-def test_task_config_with_retry(mock_retrying_executor):
-    mock_task_config = _get_mock_task_config()
+def test_retry_loop_retries_task(mock_retrying_executor, mock_event):
+    mock_event = mock_event.set('terminal', True)
+    mock_retrying_executor.stopping = True
+    mock_retrying_executor._is_current_attempt = mock.Mock(return_value=True)
+    mock_retrying_executor._restore_task_id = mock.Mock(
+        return_value=mock_event)
+    mock_retrying_executor.retry = mock.Mock(return_value=True)
+    mock_retrying_executor.retry_pred = mock.Mock(return_value=True)
+    mock_retrying_executor.src_queue = Queue()
+    mock_retrying_executor.src_queue.put(mock_event)
+    mock_retrying_executor.task_retries = mock_retrying_executor.\
+        task_retries.set(mock_event.task_id, 1)
+
+    mock_retrying_executor.retry_loop()
+
+    assert mock_retrying_executor.dest_queue.qsize() == 0
+    assert mock_retrying_executor.retry.call_count == 1
+
+
+def test_retry_loop_does_not_retry_task(mock_retrying_executor, mock_event):
+    mock_event = mock_event.set('terminal', True)
+    mock_retrying_executor.stopping = True
+    mock_retrying_executor._is_current_attempt = mock.Mock(return_value=True)
+    mock_retrying_executor.retry = mock.Mock(return_value=False)
+    mock_retrying_executor.retry_pred = mock.Mock(return_value=False)
+    mock_retrying_executor.task_retries = mock_retrying_executor.\
+        task_retries.set(mock_event.task_id, 1)
+    modified_task_id = mock_event.task_id + '-retry1'
+    modified_mock_event = mock_event.set(
+        'task_id',
+        modified_task_id
+    )
+    mock_retrying_executor.src_queue = Queue()
+    mock_retrying_executor.src_queue.put(modified_mock_event)
+
+    mock_retrying_executor.retry_loop()
+
+    assert mock_retrying_executor.dest_queue.qsize() == 1
+    assert len(mock_retrying_executor.task_retries) == 0
+
+
+def test_retry_loop_filters_out_non_task(mock_retrying_executor):
+    mock_event = Event(
+        kind='control',
+        raw='some message',
+        message='stop',
+        terminal=True
+    )
+
+    mock_retrying_executor.stopping = True
+    mock_retrying_executor._is_current_attempt = mock.Mock(return_value=True)
+    mock_retrying_executor.event_with_retries = mock.Mock()
+    mock_retrying_executor.retry = mock.Mock()
+    mock_retrying_executor.src_queue = Queue()
+    mock_retrying_executor.src_queue.put(mock_event)
+
+    mock_retrying_executor.retry_loop()
+
+    assert mock_retrying_executor.dest_queue.qsize() == 1
+    assert mock_retrying_executor.retry.call_count == 0
```

Comment:
I think it's a little overkill to assert this, the code would definitely break before then because control events don't have task ids, etc. It's enough to know that the task got passed on to dest_queue and the loop didn't raise an exception. That way you don't need as much setup code, which would need updating in the future.

Reply:
Fair.